### ISSUE TITLE
Feature Request: Optional dependency version specifier

### ISSUE THREAD

**Issue by schneems** [rubygems/rubygems#1811]
I would like to suggest a feature.

My current problem is when developing a gem with "optional" dependencies there is no way to tell rubygems a valid gem version specification.

For example in Sprockets we have optional code that can be run if you have `uglifier` gem on your system. However not all users of Sprockets wish to use Uglifier so we do not want to depend on it in the gemspec. Recently there was a breaking change in the Uglifier API, if we were directly depending on this gem then I would solve the problem by only allowing users with uglifier >= 3.0.4 with Sprockets 4. However since we are not declaring this as a dependency, there is no way to do this.

I am proposing a declaration in the Gemspec like this:

```
  s.optional_dependency "uglifier", ">= 3.0.4"
```

Proposed behavior: If this was in a gemspec then if the Sprockets gem is loaded first, then when the Uglifier gem will be subject to the additional version specification. If the Uglifier gem is loaded first then the system would error out while trying to load Sprockets if it is not within the specification. Ideally if bundler could eventually integrate to resolve optional dependencies at install time, that would be 👌.

Current behavior: Right now the only way to handle the case is to branch as in this PR https://github.com/rails/sprockets/pull/448 or to manually detect versions at runtime and error out. Proposed behavior would make the case where sprockets is required first much better, while the other case of uglifier loaded first would be essentially the same (loadtime/runtime error). Eventual ability to coordinate with bundler would be the ultimate end goal. 

I see this as beneficial for a wide range of gems:

- multi_json: supports many json backends but does not depend on ALL of them being loaded to work, just one.
- active_record: supports many database backends but does not depend on ALL of them being loaded to work, just one.
- Sprockets as mentioned
- The sentry client `raven` maintainer mentioned they have optional gems.
- etc.

This is a pretty significant feature proposal/request. I'm curious what you think of the overall idea as well as any currently available alternatives that don't require a feature in rubygems that I've not already touched on.

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by segiddins** [comment]
I actually already implemented support for this in the resolver, but the major hurdles are backwards compatibility and integration with the multiple index formats currently used by rubygems and bundler.

**Comment by indirect** [comment]
@schneems thanks for opening this request! this is a feature that would definitely be helpful for a pretty wide range of gems (pry and pry plugins comes to mind for me, along with the other examples you mentioned).

RubyGems is just now getting to the point where we no longer have any embarrassing multiple-year-old bugs left to handle, and this is something that I would like to do in the future. As @segiddins mentioned, the technical underpinnings are already complete, and all that's left is the actual hard part: creating an API for this that makes sense and is usable, creating a working implementation, and ironing out all of the inevitable edge cases.

The last project this big that we shipped was the compact index, and that took 3.5 years from the time I proposed a spec to the day that we shipped a version of Bundler that used it by default. My personal wild guesstimate for this feature, which would require changes across RubyGems, Bundler, and RubyGems.org, is approximately 3 months of full-time work from a single dev who is already familiar with all of those projects (vaguely around 500 hours). Today, Ruby Together can afford to pay for about 5 hours of work per week on each one of those three projects, and almost all of that time is currently dedicated to needed triage, bugfixes, and shipping smaller features.

If more companies start contributing to Ruby Together and we can start paying for more hours, I think this feature could feasibly be implemented sometime in the next calendar year. If the companies that use RubyGems to make money continue to pass on contributing back to RubyGems infra, this feature will stay on our list of "clearly a good idea that we should work on when we have time". At that point, I suspect a reasonable timeline is more like the 3-4 years it took to finish the compact index.

**Comment by drbrain** [comment]
To further justify the large time estimate

> Proposed behavior:
> If [`sprockets.optional_dependency "uglifier", "~> 3.0"`] was in a gemspec then if the Sprockets gem is loaded first, then when the Uglifier gem will be subject to the additional version specification.

I foresee a problem:

I read the sprockets documentation.

I see that requiring sprockets and uglifier together will give me an extra feature.

I want this extra feature.

I type in my program `require 'uglifier'` above `require 'sprockets'` because my editor starts me on the first line of my program.

Expected:

I get the special sprockets-uglifier optional features

Actual:

I don't get the sprockets-uglifier optional features because my require order is wrong.

I believe this would be a source of complaints that we (RubyGems) would have to deal with because RubyGems behaves differently because you put your `require` in a different order than someone else.

> If the Uglifier gem is loaded first then the system would error out while trying to load Sprockets if it is not within the specification.

I foresee a problem:

I start using `sprockets` and everything is great.

For some completely unrelated reason I add `uglifier`, but I choose a version that is not compatible with the sprockets optional dependency.

I don't know about the sprockets-uglifier optional dependency.

I don't want to use the sprockets-uglifier optional features.

Expected:

My program works.

Actual:

My program refuses to run because what was listed as an "optional" dependency is now being enforced.

I believe this would be a source of complaints that we (RubyGems) would have to deal with because some users won't want to use the optional feature.

Please note that the expectation for the first problem (feature 1 + feature 2 = 4 features) conflicts with the expectation for the second problem (feature 1 + feature 2 = 3 features)

Optional dependencies have a large set of behaviors and expectations that need to be set before work can start.  Defining what an "optional" dependency means for RubyGems users will take a significant amount of time.  I think @indirect's estimate above may be optimistic.  

Given the large (implying risky) time estimate above is there some other thing RubyGems can add that addresses the root issue (automatic dependency checks) that doesn't involve answering so many questions about behavior?

Maybe improved API for gems that use gems they don't directly depend on?

Maybe documentation that recommends how to use the existing API for this use case?

**Comment by chrismo** [comment]
`s.optional_dependency "uglifier", ">= 3.0.4"`

> Proposed behavior: If this was in a gemspec then ... If the Uglifier gem is loaded first then the system would error out while trying to load Sprockets if it is not within the specification.

Without thinking about it too much, that doesn't make sense to me. I'd think, "but it says it's optional - why is it erring out?" My knee-jerk presumption is if I already have an uglifier version that doesn't match, but you (sprockets) say it's optional, then why are you blocking me?

/my half-cent

**Comment by chrismo** [comment]
i also wonder if this would be too hard (it could be) to do as a plugin. if not, plugin is a great way to experiment with it all.

**Comment by schneems** [comment]
I'm open on the name, that's a bit of a bikeshed at this point, could eastilly do something else though, `s.dependency_constraint_if_present "uglifier", ">= 3.0.4"`

**Comment by chrismo** [comment]
that's more intentional revealing to me, though a bit uglier. 

i've been through all of this recently adding in 'conservative' updates to bundle update. Concise, intention-revealing names are really hard. so - i don't consider that much of a bike shed, fwiw. that may be in line with @drbrain's comments too - that there are different ways to interpret 'optional'?

**Comment by chrismo** [comment]
I'm not sure `dependency_constraint_if_present` is what @segiddins implemented in Molinillo either. Following through the CHANGELOG here: https://github.com/CocoaPods/Molinillo/blob/master/CHANGELOG.md#030-2015-06-29

to:  https://github.com/CocoaPods/Molinillo/blob/master/lib/molinillo/modules/specification_provider.rb#L91-L98

But could be I'm looking in the wrong place.

**Comment by segiddins** [comment]
@chrismo optional / "weak" dependencies haven't been merged yet

**Comment by chrismo** [comment]
ah - this branch?  https://github.com/CocoaPods/Molinillo/compare/seg-optional-dependencies#diff-b830bdc867335e9e433851661bd04fb9R292

**Comment by kbrock** [comment]
This would be a great feature.
For some background around this, you may also want to look through #1404 #346

**Comment by indirect** [comment]
@schneems it doesn't seem particularly good to ship a new feature that allows sprockets to dictate what versions of uglify can be installed at the same time, even for users who actively don't want the sprockets-uglify functionality. This is (more or less) why one traditional solution to this problem has been a gem named `sprockets-uglify` that depends on both gems and activates the relevant functionality. Then people who want it get it, and people who don't want it don't have their (unrelated) copy of the uglify gem messed with.

Right now, at least, I'm not comfortable with this feature unless it is explicitly opt-in on the part of the end user dev. Just adding sprockets and uglify to the same gemfile doesn't mean that optional deps are intended to be activated.

**Comment by schneems** [comment]
All good points. 

The `sprockets-uglifier` approach might be fine if there weren't a matrix of optional dependencies 

```
  s.add_development_dependency "babel-transpiler", "~> 0.6"
  s.add_development_dependency "closure-compiler", "~> 1.1"
  s.add_development_dependency "coffee-script-source", "~> 1.6"
  s.add_development_dependency "coffee-script", "~> 2.2"
  s.add_development_dependency "eco", "~> 1.0"
  s.add_development_dependency "ejs", "~> 1.0"
  s.add_development_dependency "execjs", "~> 2.0"
  s.add_development_dependency "jsminc", "~> 1.1"
  s.add_development_dependency "minitest", "~> 5.0"
  s.add_development_dependency "nokogiri", "~> 1.3"
  s.add_development_dependency "rack-test", "~> 0.6"
  s.add_development_dependency "rake", "~> 10.0"
  s.add_development_dependency "sass", "~> 3.4"
  s.add_development_dependency "sassc", ">= 1.10.1", "< 2.0"
  s.add_development_dependency "uglifier", "~> 2.3"
  s.add_development_dependency "yui-compressor", "~> 0.12"
  s.add_development_dependency "zopfli", "~> 0.0.4"
```

I would need to have `sprockets-uglifier-zopfli-bable-transpiler-sassc` and everything in-between.

I'm wondering if the right approach for sprockets might be to have more dependencies but still only autoload them if they're being used. However this would still have the same issue you're referring to, where someone wants to use uglifier but not with sprockets, either way sprockets would have to impose limitations on the dependencies at install time with incomplete information over what the user wants. I think whether the dev used a `dependency_constraint_if_present` or `add_dependency` then either way they are hijacking a user's choice, the primary difference is that in one scenario `uglifier` would be installed even if you didn't need it, and not in the proposed scenario.

**Comment by indirect** [comment]
@schneems yup. that point is what makes me think this feature will require both an extension to the gemspec and an extension to the gen syntax in the gemfile. Maybe something
like:

```
gem "sprockets", include: [:uglify, :sassc]
```

Then, and only then, would the optional dependency requirements be enabled.

How does that sound?

**Comment by segiddins** [comment]
We went that route in CocoaPods (subspecs) and supporting them has been .... difficult. Too many edge cases.

**Comment by indirect** [comment]
@segiddins what is "that route"? how do subspecs work?

**Comment by segiddins** [comment]
The route of specs being able to specify "optional features". https://guides.cocoapods.org/syntax/podspec.html#subspec

**Comment by indirect** [comment]
I see. That’s... pretty different from what we’re talking about here. Optional deps would not be contained inside the gems that depend on them, and we are not going to be allowing other gems to depend on a specific optional dependency being activated.

**Comment by matthewd** [comment]
This feels like it could overlap with the oft-requested ability for a Gemfile to say "I know that this gem A has a restrictive dependency on gem B. I am smarter than the author of A.gemspec, so please ignore that particular constraint."

Here that would be

```ruby
gem "sprockets", disregard_dependency: ["uglifier"]
```

but it could also be

```ruby
gem "rack", "4.0"

# this gem doesn't realise it, but I know it's actually compatible with the newer version
# (or similarly, I want to run like this to *find out*)
gem "actionpack", disregard_dependency: ["rack"]
```

**Comment by schneems** [comment]
I like the opt out behavior proposed by @matthewd. It would allow a gem author to write for what the believe to be the 80% case while giving an easy out to anyone else.

**Comment by indirect** [comment]
As we have discussed at extreme length elsewhere, we are not going to allow monkey-patching gemspecs via the Gemfile.

I’m honestly still pretty skeptical of even something like `gem “sprockets”, without: [:uglifier]` because it does nothing for the options that a gem author wants to allow but thinks should be off by default. So then what do you do? Add both include: and without:? That sounds like a nightmare of edge cases. Please no. :)

**Comment by shevegen** [comment]
I agree with the proposal - I also have some gems where I have optional dependencies.

It would be nice to specify these; in particular where ruby gems could continue to download
and make use of these, without refusing to continue. Some gems are like optional add-ons.

To the comment about lack of manpower - I can understand this very well; I also see it in
the rubygems.org issue request tracker where some features get implemented and others
have to wait for some longer time.

I believe that the optional dependency here, though, would be a fairly useful addition and 
would hopefully be eventually prioritized. 

PS: Note that I refer to the original suggestion here, via the

  **s.optional_dependency "uglifier", ">= 3.0.4",**

example.

I am indifferent to Gemfile as I do not use the latter, I only use .gemspec files,
so parts such as 

  **gem "actionpack", disregard_dependency: ["rack"]**

are not the scope of my +1 given there, simply because I only use .gemspec
and not Gemfile so I can not really comment on what OTHER people use
or find useful there. (On a side note, my gems currently all have dependency
for my own, other gems, as version 0 or greater, largely because I have no
easy way to declare that some of my gems are optional; and also because 
failures at before-install time are more annoying to me than failures during
run-time, but this is another issue - just a side comment from me here.)

**Comment by drbrain** [comment]
@shevegen The main issue is behavior and user expectation when optional dependencies are present.

### How do you expect optional dependencies to behave with respect to incompatible versions?

In [my comment above](https://github.com/rubygems/rubygems/issues/1811#issuecomment-270799286) what should happen if a user wants to use sprockets and uglify, but wants to use a version that isn't compatible with the optional dependency because they don't want those features?

### How do users use the features provided by the optional dependency?

An optional dependency implies additional features for a library.  Are they enabled automatically when the optional dependency is present?  Is there new RubyGems API that enables them?

How does the user discover them?  Is there install-time messaging?  New CLI commands?

Without deciding what "optional dependency" means we can't implement it at all.

**Comment by BookOfGreg** [comment]
I'm reading this thread due to my work on reactjs/react-rails

In react-rails 2.2, if Webpacker is defined it will use that to load the assets, and v2.2 supports Webpacker 1.x and 2.0 but not 3.0.
After reactjs/react-rails#784 It will support Webpacker 2.0 and 3.x but not 1.x anymore.

I would expect the behaviour to be, if Webpacker is in the gemfile and react-rails is 2.2, it has an additional dependency of **'webpacker', '<3'**, and if webpacker 1 is dropped then the dependency becomes **'>=2', '<4'** and otherwise resolves as normal.

That way if someone has Webpacker pegged to version 1.x, it would resolve to react-rails 2.2 or lower.

This would make the semantics "If in the gemfile, use the existing dependency logic"

**Comment by AlexEvanczuk** [comment]
Hi everyone, thanks for this interesting and vibrant discussion. Hi @indirect! Hope all is well for you.

I wanted to present another example of a use case that our organization has for this.

We are using [`sorbet`](https://sorbet.org/) (Stripe's type checker for Ruby) in some of our gems, but we would like clients of those gems to not have to include `sorbet-runtime` as a dependency.

Instead, we want to load a separate gem, which is a shim for `sorbet-runtime`, if the client does not have `sorbet-runtime` installed.

See https://github.com/rubygems/rubygems/issues/3607 for my original issue and question.

My ideal syntax is something like this: 
```
spec.add_optional_dependency 'sorbet-runtime', '~> 0.5.5585', else_add: 'sorbet-runtime-stub', '~> 0.14'
```

For me this has the following behaviors:
1. In the case that the client includes `sorbet-runtime` in its `Gemfile`, then we ensure the versions are compatible, and error as normal if not.
2. In the case the client does not include `sorbet-runtime`, then we fall back to loading the runtime stub.

I ended up trying to implement a version of this behavior when the code loads: https://github.com/rubygems/rubygems/issues/3607#issuecomment-626088347

**Comment by kbrock** [comment]
Hi @AlexEvanczuk 

It sounds to me that you need the shim unless at runtime `sorbert-runtime` is available.
I'd make `sorbert-shim` a dependency, because that has to be around, and then at runtime I'd load `sorbert-runtime` and rescue `LoadError` and require `sorbert-shim`.

**Comment by AlexEvanczuk** [comment]
> Hi @AlexEvanczuk
> 
> It sounds to me that you need the shim unless at runtime `sorbert-runtime` is available.
> I'd make `sorbert-shim` a dependency, because that has to be around, and then at runtime I'd load `sorbert-runtime` and rescue `LoadError` and require `sorbert-shim`.
> 
> it is not perfect, but, to me, it seems the simplest solution to guarantee that at least the bare minimum is required.

Thanks @kbrock -- I linked a comment above where I did exactly that.

However, this approach has a lot of downsides that bundler can solve.  Namely, this approach is sort of "lying" to the client, since the client typically expects the `Gemfile.lock` to comprehensively describe the 3rd party dependencies of a library. If we had the optional dependency feature, then either `sorbet-runtime-stub` or `sorbet-runtime` would be in `Gemfile.lock`, but not either. I think this would be a huge boon for the dependency footprint of this gem given this use case!

**Comment by simi** [comment]
1. I think this should be moved to https://github.com/rubygems/rfcs/issues
2. As mentioned, there are already ways how to use "optional" dependencies (even with version contraints) in runtime+ for example with `gem` + `rescue Gem::LoadError` combo. I think good start would be to contribute that to RubyGems guides.

**Comment by AlexEvanczuk** [comment]
@simi 
I created an RFC there: https://github.com/rubygems/rfcs/issues/27

In that RFC, I listed some issues with (2). I'm concerned with those issues enough that I'm not sure it would be best to recommend that -- I listed some of those reasons in the RFC.

**Comment by kbrock** [comment]
I think @alexevanczuk posted a good proposal, and having the consumer of the optional dependency opt-in, I think it resolves all the concerns around having an optional dependency forced upon you.

Has anyone checked out the RFC? https://github.com/rubygems/rfcs/issues/27

**Comment by kbrock** [comment]
I thought the proposal had a nice opt-in syntax.

```
# gemspec for the gem
spec.add_optional_dependency 'sorbet-runtime', '~> 0.5.5585'
spec.add_optional_dependency 'sorbet-runtime-stub', '~> 0.14'

# gemfile for your project
gem 'my_special_gem', include_optional_dependencies: 'sorbet-runtime'
gem 'activerecord', include_optional_dependencies: 'sqlite'
```

This doesn't get in my way. It is offered and available if I ask for it via `include_optional_dependencies: "sqlite"`.

My issue is the optional dependencies are hidden and not declared in the gemspec or on rubygems.org. It blows up or doesn't work at runtime.
